In [ ]:
import pandas as pd
#you can find the data here https://drive.google.com/drive/folders/1FFo_aITk4hq3y2HWPHOubfnyDil5djzS?usp=sharing

In [ ]:
df_high=pd.read_csv("/content/drive/MyDrive/Parsed_Files/Aggregative/high_quality_monthly.csv")
df_mid=pd.read_csv("/content/drive/MyDrive/Parsed_Files/Aggregative/medium_quality_monthly.csv")
df_low=pd.read_csv("/content/drive/MyDrive/Parsed_Files/Aggregative/low_quality_monthly.csv")
df_overall=pd.read_csv("/content/drive/MyDrive/Parsed_Files/Aggregative/overall_monthly.csv")

In [ ]:
type(df_overall['year'][0])

numpy.int64

In [ ]:
df_overall = pd.merge(df_overall, df_high,on=['year', 'month', 'State'])
df_overall = pd.merge(df_overall, df_low,on=['year', 'month', 'State'])
df_overall = pd.merge(df_overall, df_mid,on=['year', 'month', 'State'])

In [ ]:
df_overall=df_overall.fillna(0)
df_overall=df_overall.replace(to_replace=0, method='ffill')

In [ ]:
df=df_overall
#df['date'] = pd.to_datetime(df['date'])
#df['year'] = df['date'].dt.year
#df['month'] = df['date'].dt.month
df = df[df['State'] != 'Vermont']
df = df[df['State'] != 'Rhode-Island']
df = df[df['State'] != 'Nebraska']
df = df[df['State'] != 'Wyoming']

df = df[(df['year'] >= 2013) & (df['year'] <= 2017)]

df.rename(columns = {'State':'state'}, inplace = True)
df

month  year  overall_amount_spent  overall_grams  \
9       January  2013                  1481          221.0   
10     February  2013                  1600          276.5   
11        March  2013                  2245          330.0   
12        April  2013                  1925          285.0   
13          May  2013                  1945          225.0   
...         ...   ...                   ...            ...   
4785     August  2017                  4320          378.5   
4786  September  2017                  1192           90.0   
4787    October  2017                  2912          380.5   
4788   November  2017                  1802          172.0   
4789   December  2017                  2223          240.0   

      overall_price_per_gram      state  high_amount_spent  high_grams  \
9                  10.109524    Alabama               1201       124.0   
10                  7.309524    Alabama                505        94.5   
11                  7.592857    Alabama               1985       236.0   
12                  8.916667    Alabama               1510       173.0   
13                 11.630952    Alabama               1110        69.5   
...                      ...        ...                ...         ...   
4785               11.677019  Wisconsin               2765       213.5   
4786               13.203297  Wisconsin               1080        77.0   
4787               11.004658  Wisconsin               1976       231.0   
4788               13.667857  Wisconsin               1617       140.0   
4789               10.972689  Wisconsin               1420       147.0   

      high_price_per_gram  low_amount_spent  low_grams  low_price_per_gram  \
9               12.484127               135       12.0           12.500000   
10               8.750000               100       28.0            3.571429   
11               9.803571               100       28.0            3.571429   
12              11.285714               100       28.0            3.571429   
13              17.040816                10        3.5            2.857143   
...                   ...               ...        ...                 ...   
4785            14.206349                60        3.5           17.142857   
4786            14.226190                60        3.5           17.142857   
4787             8.376984                60        3.5           17.142857   
4788            11.550000                60        3.5           17.142857   
4789            10.816327                60        3.5           17.142857   

      medium_amount_spent  medium_grams  medium_price_per_gram  
9                     280          97.0               6.547619  
10                    995         154.0               6.924603  
11                    260          94.0               3.171429  
12                    415         112.0               4.178571  
13                    825         152.0               7.474490  
...                   ...           ...                    ...  
4785                 1555         165.0              10.051020  
4786                  112          13.0              12.326531  
4787                  936         149.5              12.693878  
4788                  185          32.0              15.785714  
4789                  803          93.0              11.082143  

[2760 rows x 15 columns]

In [ ]:
for i in set(df["state"].tolist()):
  print(len(df[df["state"]==i]),i)

60 Colorado
60 Pennsylvania
60 Minnesota
60 South-Dakota
60 California
60 New-Mexico
60 Georgia
60 Massachusetts
60 Connecticut
60 Nevada
60 West-Virginia
60 Arizona
60 Wisconsin
60 Florida
60 Alabama
60 Maine
60 Missouri
60 Alaska
60 District-of-Columbia
60 North-Dakota
60 Illinois
60 Oklahoma
60 Indiana
60 Arkansas
60 Ohio
60 Tennessee
60 Michigan
60 Maryland
60 South-Carolina
60 Delaware
60 Kansas
60 Montana
60 North-Carolina
60 Utah
60 Washington
60 Hawaii
60 Iowa
60 New-Hampshire
60 Virginia
60 Mississippi
60 Oregon
60 Idaho
60 Texas
60 New-Jersey
60 Louisiana
60 Kentucky


In [ ]:
df.to_csv("mugu.csv")

In [ ]:
deaths=pd.read_csv("/content/drive/MyDrive/Parsed_Files/Aggregative/temp.csv")
#month_dict = {'January': 1, 'February': 2, 'March': 3, 'April': 4, 'May': 5, 'June': 6,'July': 7, 'August': 8, 'September': 9, 'October': 10, 'November': 11, 'December': 12}
#deaths['month'] = deaths['month'].replace(month_dict)
deaths = deaths[deaths['state'] != 'Vermont']
deaths = deaths[deaths['state'] != 'Rhode-Island']
deaths = deaths[deaths['state'] != 'Nebraska']
deaths = deaths[deaths['state'] != 'Wyoming']

In [ ]:
deaths

month  year          state  deaths  injured
0         April  2013        Alabama       0        0
1         April  2013         Alaska       0        0
2         April  2013        Arizona       2        4
3         April  2013       Arkansas       0        0
4         April  2013     California       1        7
...         ...   ...            ...     ...      ...
2934  September  2017           Utah       2       10
2935  September  2017       Virginia      34       79
2936  September  2017     Washington      24       34
2937  September  2017  West-Virginia       5        5
2938  September  2017      Wisconsin      20       55

[2760 rows x 5 columns]

In [ ]:
states1=set(deaths["state"].tolist())
states2=set(df["state"].tolist())

In [ ]:
if(states1==states2):
  print("Great!")

Great!


In [ ]:
merged_df = pd.merge(df, deaths, on=['year', 'month', 'state'])

In [ ]:
merged_df

month  year  overall_amount_spent  overall_grams  \
0       January  2013                  1481          221.0   
1      February  2013                  1600          276.5   
2         March  2013                  2245          330.0   
3         April  2013                  1925          285.0   
4           May  2013                  1945          225.0   
...         ...   ...                   ...            ...   
2755     August  2017                  4320          378.5   
2756  September  2017                  1192           90.0   
2757    October  2017                  2912          380.5   
2758   November  2017                  1802          172.0   
2759   December  2017                  2223          240.0   

      overall_price_per_gram      state  high_amount_spent  high_grams  \
0                  10.109524    Alabama               1201       124.0   
1                   7.309524    Alabama                505        94.5   
2                   7.592857    Alabama               1985       236.0   
3                   8.916667    Alabama               1510       173.0   
4                  11.630952    Alabama               1110        69.5   
...                      ...        ...                ...         ...   
2755               11.677019  Wisconsin               2765       213.5   
2756               13.203297  Wisconsin               1080        77.0   
2757               11.004658  Wisconsin               1976       231.0   
2758               13.667857  Wisconsin               1617       140.0   
2759               10.972689  Wisconsin               1420       147.0   

      high_price_per_gram  low_amount_spent  low_grams  low_price_per_gram  \
0               12.484127               135       12.0           12.500000   
1                8.750000               100       28.0            3.571429   
2                9.803571               100       28.0            3.571429   
3               11.285714               100       28.0            3.571429   
4               17.040816                10        3.5            2.857143   
...                   ...               ...        ...                 ...   
2755            14.206349                60        3.5           17.142857   
2756            14.226190                60        3.5           17.142857   
2757             8.376984                60        3.5           17.142857   
2758            11.550000                60        3.5           17.142857   
2759            10.816327                60        3.5           17.142857   

      medium_amount_spent  medium_grams  medium_price_per_gram  deaths  \
0                     280          97.0               6.547619       0   
1                     995         154.0               6.924603       0   
2                     260          94.0               3.171429       0   
3                     415         112.0               4.178571       0   
4                     825         152.0               7.474490       0   
...                   ...           ...                    ...     ...   
2755                 1555         165.0              10.051020      18   
2756                  112          13.0              12.326531      20   
2757                  936         149.5              12.693878      17   
2758                  185          32.0              15.785714      15   
2759                  803          93.0              11.082143      15   

      injured  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
2755       56  
2756       55  
2757       45  
2758       57  
2759       19  

[2760 rows x 17 columns]

In [ ]:
merged_df.corr()

year  overall_amount_spent  overall_grams  \
year                    1.000000             -0.189822      -0.269061   
overall_amount_spent   -0.189822              1.000000       0.859260   
overall_grams          -0.269061              0.859260       1.000000   
overall_price_per_gram  0.099280              0.147578      -0.106303   
high_amount_spent      -0.121348              0.848467       0.721071   
high_grams             -0.119660              0.768234       0.829583   
high_price_per_gram     0.065108              0.116091      -0.098987   
low_amount_spent        0.056071              0.106500      -0.017543   
low_grams              -0.084719              0.055968       0.093045   
low_price_per_gram      0.052542              0.069419      -0.015079   
medium_amount_spent    -0.157454              0.724352       0.644976   
medium_grams           -0.292705              0.637918       0.812326   
medium_price_per_gram   0.110161              0.088359      -0.112594   
deaths                  0.329521              0.028135       0.022692   
injured                 0.318378              0.030242      -0.003638   

                        overall_price_per_gram  high_amount_spent  high_grams  \
year                                  0.099280          -0.121348   -0.119660   
overall_amount_spent                  0.147578           0.848467    0.768234   
overall_grams                        -0.106303           0.721071    0.829583   
overall_price_per_gram                1.000000           0.060853   -0.066804   
high_amount_spent                     0.060853           1.000000    0.888176   
high_grams                           -0.066804           0.888176    1.000000   
high_price_per_gram                   0.525522           0.134555   -0.083173   
low_amount_spent                      0.287748           0.039840    0.000267   
low_grams                             0.000244           0.031836    0.039997   
low_price_per_gram                    0.280177           0.009780   -0.004550   
medium_amount_spent                   0.113756           0.331803    0.322609   
medium_grams                         -0.114427           0.337356    0.411008   
medium_price_per_gram                 0.621591           0.017013   -0.060974   
deaths                               -0.006810           0.022410    0.020394   
injured                               0.008643           0.019345   -0.005151   

                        high_price_per_gram  low_amount_spent  low_grams  \
year                               0.065108          0.056071  -0.084719   
overall_amount_spent               0.116091          0.106500   0.055968   
overall_grams                     -0.098987         -0.017543   0.093045   
overall_price_per_gram             0.525522          0.287748   0.000244   
high_amount_spent                  0.134555          0.039840   0.031836   
high_grams                        -0.083173          0.000267   0.039997   
high_price_per_gram                1.000000          0.064949   0.006404   
low_amount_spent                   0.064949          1.000000   0.339816   
low_grams                          0.006404          0.339816   1.000000   
low_price_per_gram                 0.011683          0.643233  -0.085934   
medium_amount_spent                0.051317          0.062278   0.033796   
medium_grams                      -0.084749         -0.031426   0.043443   
medium_price_per_gram              0.254479          0.144280  -0.013275   
deaths                             0.015987         -0.080953  -0.039088   
injured                            0.030734         -0.080951  -0.061435   

                        low_price_per_gram  medium_amount_spent  medium_grams  \
year                              0.052542            -0.157454     -0.292705   
overall_amount_spent              0.069419             0.724352      0.637918   
overall_grams                    -0.015079             0.644976      0.812326   
overall_price_per_gra

In [ ]:
def calculate_difference_list(lst):
    diff_lst = []
    for i in range(len(lst)-1):
        diff_lst.append(lst[i+1] - lst[i])
    return diff_lst

alabama=merged_df[merged_df["state"]=="Alabama"]
listx=calculate_difference_list(alabama["overall_amount_spent"].tolist())
listy=calculate_difference_list(alabama["deaths"].tolist())

In [ ]:
import statsmodels.api as sm

ols_fit_overall=pd.DataFrame()
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
for i in set(merged_df["state"]):
  states_df=merged_df[merged_df["state"]==i]
  listx=calculate_difference_list(states_df["overall_amount_spent"].tolist())
  listy=calculate_difference_list(states_df["deaths"].tolist())
  listx = sm.add_constant(listx)
  ols = sm.OLS(listy,listx,missing = 'drop')
  ols_result = ols.fit()
  ols_new_results=ols_result.get_robustcov_results()
  print(ols_new_results.summary())
  states.append(i)
  rsq.append(ols_new_results.rsquared)
  adjrsq.append(ols_new_results.rsquared_adj)
  pval.append(ols_new_results.pvalues[1])
  tval.append(ols_new_results.tvalues[1])

ols_fit_overall["State"]=states
ols_fit_overall["R^2"]=rsq
ols_fit_overall["Adj_R^2"]=adjrsq
ols_fit_overall["pval"]=pval
ols_fit_overall["tval"]=tval

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                     1.446
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.234
Time:                        03:58:42   Log-Likelihood:                -200.07
No. Observations:                  59   AIC:                             404.1
Df Residuals:                      57   BIC:                             408.3
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2209      0.951      0.232      0.8

In [ ]:
ols_fit_overall.sort_values(by="R^2", ascending = False).to_markdown()

'|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |\n|---:|:---------------------|------------:|-------------:|------------:|-----------:|\n| 13 | Florida              | 0.0888685   |  0.0728837   | 0.000865588 | -3.51644   |\n| 44 | Louisiana            | 0.0638775   |  0.0474543   | 0.0573581   | -1.9398    |\n| 30 | Kansas               | 0.062849    |  0.0464077   | 0.0826275   |  1.76673   |\n| 15 | Maine                | 0.0562343   |  0.039677    | 0.0537864   | -1.9693    |\n|  1 | Pennsylvania         | 0.0527094   |  0.0360903   | 0.0672851   | -1.86533   |\n| 18 | District-of-Columbia | 0.0461846   |  0.029451    | 0.119369    |  1.58119   |\n| 33 | Utah                 | 0.0416968   |  0.0248844   | 0.12728     | -1.54749   |\n| 42 | Texas                | 0.0367197   |  0.01982     | 0.154155    |  1.44421   |\n| 19 | North-Dakota         | 0.0319615   |  0.0149784   | 0.201515    |  1.29216   |\n|  4 | California           | 0.0306821   

|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |
|---:|:---------------------|------------:|-------------:|------------:|-----------:|
| 42 | Florida              | 0.0888685   |  0.0728837   | 0.000865588 | -3.51644   |
| 12 | Louisiana            | 0.0638775   |  0.0474543   | 0.0573581   | -1.9398    |
|  6 | Kansas               | 0.062849    |  0.0464077   | 0.0826275   |  1.76673   |
| 27 | Maine                | 0.0562343   |  0.039677    | 0.0537864   | -1.9693    |
| 23 | Pennsylvania         | 0.0527094   |  0.0360903   | 0.0672851   | -1.86533   |
| 20 | District-of-Columbia | 0.0461846   |  0.029451    | 0.119369    |  1.58119   |
| 10 | Utah                 | 0.0416968   |  0.0248844   | 0.12728     | -1.54749   |
| 24 | Texas                | 0.0367197   |  0.01982     | 0.154155    |  1.44421   |
| 26 | North-Dakota         | 0.0319615   |  0.0149784   | 0.201515    |  1.29216   |
| 18 | California           | 0.0306821   |  0.0136765   | 0.084289    |  1.75699   |
| 44 | Missouri             | 0.0284647   |  0.0114203   | 0.157312    | -1.43303   |
| 22 | Hawaii               | 0.0262477   |  0.00916433  | 0.416022    | -0.81931   |
| 40 | North-Carolina       | 0.0252741   |  0.00817365  | 0.207711    | -1.27435   |
| 33 | Montana              | 0.0245803   |  0.00746768  | 0.00536594  | -2.895     |
|  0 | Maryland             | 0.0241135   |  0.00699266  | 0.218604    | -1.24397   |
|  3 | Alaska               | 0.0210938   |  0.00391997  | 0.27902     |  1.09294   |
| 14 | Washington           | 0.0197866   |  0.00258992  | 0.185526    | -1.34011   |
| 25 | Virginia             | 0.0196821   |  0.00248355  | 0.225476    | -1.22537   |
| 32 | West-Virginia        | 0.0180827   |  0.000856082 | 0.224819    |  1.22713   |
| 39 | Ohio                 | 0.0171156   | -0.000127944 | 0.33127     | -0.979914  |
|  9 | Colorado             | 0.0158626   | -0.00140301  | 0.234101    | -1.20261   |
| 13 | Idaho                | 0.0144846   | -0.00280517  | 0.149006    | -1.46283   |
|  8 | Connecticut          | 0.013342    | -0.00396784  | 0.328539    | -0.985509  |
| 43 | Georgia              | 0.0130802   | -0.00423422  | 0.319291    | -1.00469   |
| 21 | Arizona              | 0.0124495   | -0.0048759   | 0.259938    |  1.13786   |
|  4 | Nevada               | 0.0114171   | -0.00592642  | 0.317162    |  1.00916   |
| 34 | Michigan             | 0.0106767   | -0.00667981  | 0.472355    | -0.723461  |
| 29 | Illinois             | 0.00953005  | -0.00784661  | 0.50207     |  0.67553   |
| 37 | Indiana              | 0.00915619  | -0.00822703  | 0.286172    |  1.07665   |
|  7 | New-Jersey           | 0.0078899   | -0.00951554  | 0.344268    |  0.953689  |
| 41 | Wisconsin            | 0.00722518  | -0.0101919   | 0.556001    | -0.592286  |
|  5 | Tennessee            | 0.0072029   | -0.0102146   | 0.410188    | -0.829668  |
| 17 | New-Mexico           | 0.00598923  | -0.0114496   | 0.512345    |  0.65932   |
| 28 | Oklahoma             | 0.00510293  | -0.0123514   | 0.656118    |  0.447628  |
| 30 | Kentucky             | 0.00365414  | -0.0138256   | 0.687944    |  0.4037    |
| 31 | Minnesota            | 0.00362498  | -0.0138553   | 0.641964    | -0.467445  |
| 36 | Arkansas             | 0.00317517  | -0.014313    | 0.619722    | -0.498969  |
| 11 | Delaware             | 0.00245948  | -0.0150412   | 0.724573    |  0.354097  |
| 16 | Mississippi          | 0.00213426  | -0.0153722   | 0.659719    |  0.442615  |
| 45 | Iowa                 | 0.00204478  | -0.0154632   | 0.676008    |  0.420078  |
| 19 | New-Hampshire        | 0.00195737  | -0.0155522   | 0.668549    |  0.430371  |
|  2 | South-Carolina       | 0.00153241  | -0.0159846   | 0.750639    | -0.319336  |
| 35 | Alabama              | 0.00150252  | -0.016015    | 0.688742    | -0.402609  |
| 38 | South-Dakota         | 0.000922346 | -0.0166053   | 0.798965    | -0.25588   |
| 15 | Oregon               | 0.000603125 | -0.0169302   | 0.875275    |  0.157669  |
|  1 | Massachusetts        | 3.13682e-05 | -0.0175119   | 0.946512    | -0.0673835 |

In [ ]:
import statsmodels.api as sm

ols_fit_overall=pd.DataFrame()
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
for i in set(merged_df["state"]):
  states_df=merged_df[merged_df["state"]==i]
  listx=calculate_difference_list(states_df["overall_price_per_gram"].tolist())
  listy=calculate_difference_list(states_df["deaths"].tolist())
  listx = sm.add_constant(listx)
  ols = sm.OLS(listy,listx,missing = 'drop')
  ols_result = ols.fit()
  ols_new_results=ols_result.get_robustcov_results()
  print(ols_new_results.summary())
  states.append(i)
  rsq.append(ols_new_results.rsquared)
  adjrsq.append(ols_new_results.rsquared_adj)
  pval.append(ols_new_results.pvalues[1])
  tval.append(ols_new_results.tvalues[1])

ols_fit_overall["State"]=states
ols_fit_overall["R^2"]=rsq
ols_fit_overall["Adj_R^2"]=adjrsq
ols_fit_overall["pval"]=pval
ols_fit_overall["tval"]=tval

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.009
Model:                            OLS   Adj. R-squared:                 -0.008
Method:                 Least Squares   F-statistic:                    0.8758
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.353
Time:                        03:58:43   Log-Likelihood:                -200.27
No. Observations:                  59   AIC:                             404.5
Df Residuals:                      57   BIC:                             408.7
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2501      0.955      0.262      0.7

In [ ]:
ols_fit_overall.sort_values(by="R^2", ascending = False).to_markdown()

'|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |\n|---:|:---------------------|------------:|-------------:|------------:|-----------:|\n| 35 | Hawaii               | 0.207554    |  0.193651    | 2.58546e-05 |  4.57868   |\n|  2 | Minnesota            | 0.125573    |  0.110232    | 0.036981    | -2.13607   |\n|  9 | Nevada               | 0.104939    |  0.0892361   | 0.0687164   |  1.85537   |\n|  7 | Massachusetts        | 0.0948699   |  0.0789904   | 0.00531624  |  2.89835   |\n| 11 | Arizona              | 0.071998    |  0.0557173   | 0.0427474   | -2.07255   |\n| 14 | Alabama              | 0.0523932   |  0.0357685   | 0.0287391   | -2.24391   |\n|  6 | Georgia              | 0.0509969   |  0.0343477   | 0.00290282  | -3.11202   |\n| 15 | Maine                | 0.03966     |  0.022812    | 0.270145    | -1.11355   |\n| 40 | Oregon               | 0.0328024   |  0.015834    | 0.000574388 |  3.64789   |\n| 32 | North-Carolina       | 0.0288552   

|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |
|---:|:---------------------|------------:|-------------:|------------:|-----------:|
| 22 | Hawaii               | 0.207554    |  0.193651    | 2.58546e-05 |  4.57868   |
| 31 | Minnesota            | 0.125573    |  0.110232    | 0.036981    | -2.13607   |
|  4 | Nevada               | 0.104939    |  0.0892361   | 0.0687164   |  1.85537   |
|  1 | Massachusetts        | 0.0948699   |  0.0789904   | 0.00531624  |  2.89835   |
| 21 | Arizona              | 0.071998    |  0.0557173   | 0.0427474   | -2.07255   |
| 35 | Alabama              | 0.0523932   |  0.0357685   | 0.0287391   | -2.24391   |
| 43 | Georgia              | 0.0509969   |  0.0343477   | 0.00290282  | -3.11202   |
| 27 | Maine                | 0.03966     |  0.022812    | 0.270145    | -1.11355   |
| 15 | Oregon               | 0.0328024   |  0.015834    | 0.000574388 |  3.64789   |
| 40 | North-Carolina       | 0.0288552   |  0.0118175   | 0.0167157   | -2.46552   |
| 45 | Iowa                 | 0.0273454   |  0.0102813   | 0.214732    | -1.25463   |
| 26 | North-Dakota         | 0.0268157   |  0.00974231  | 0.000443076 |  3.72984   |
| 20 | District-of-Columbia | 0.0235431   |  0.00641223  | 0.000240661 |  3.91913   |
|  5 | Tennessee            | 0.0230968   |  0.00595814  | 0.183162    |  1.34746   |
|  3 | Alaska               | 0.0207674   |  0.00358783  | 0.212309    |  1.26138   |
| 16 | Mississippi          | 0.0181595   |  0.000934206 | 0.370869    |  0.90197   |
| 11 | Delaware             | 0.0167294   | -0.000520925 | 0.233294    | -1.20471   |
| 36 | Arkansas             | 0.0159578   | -0.00130608  | 0.142985    |  1.48526   |
|  6 | Kansas               | 0.0150687   | -0.00221084  | 0.255771    | -1.14798   |
| 19 | New-Hampshire        | 0.0135609   | -0.00374507  | 0.122615    |  1.56715   |
| 32 | West-Virginia        | 0.0127834   | -0.00453614  | 0.168354    | -1.39525   |
| 42 | Florida              | 0.01241     | -0.00491618  | 0.219249    | -1.2422    |
| 17 | New-Mexico           | 0.0113034   | -0.00604211  | 0.449955    | -0.760728  |
|  2 | South-Carolina       | 0.0100085   | -0.00735972  | 0.383824    | -0.877638  |
| 37 | Indiana              | 0.00936944  | -0.00801004  | 0.406199    |  0.836801  |
|  9 | Colorado             | 0.0091017   | -0.00828248  | 0.353305    | -0.935837  |
| 44 | Missouri             | 0.00774217  | -0.00966587  | 0.138751    | -1.50148   |
|  0 | Maryland             | 0.00510654  | -0.0123477   | 0.202272    |  1.28997   |
| 33 | Montana              | 0.00443574  | -0.0130303   | 0.474017    |  0.720737  |
|  8 | Connecticut          | 0.00383862  | -0.0136379   | 0.573111    |  0.566747  |
| 41 | Wisconsin            | 0.00332759  | -0.0141579   | 0.743853    |  0.328346  |
| 28 | Oklahoma             | 0.00327014  | -0.0142163   | 0.118434    |  1.58529   |
| 29 | Illinois             | 0.00280682  | -0.0146878   | 0.30891     | -1.02668   |
| 24 | Texas                | 0.00175567  | -0.0157574   | 0.684778    | -0.408034  |
| 10 | Utah                 | 0.00158248  | -0.0159336   | 0.5572      | -0.590484  |
| 12 | Louisiana            | 0.00144233  | -0.0160762   | 0.772528    |  0.290447  |
| 14 | Washington           | 0.00101607  | -0.01651     | 0.736716    | -0.337852  |
|  7 | New-Jersey           | 0.000973228 | -0.0165536   | 0.606413    | -0.518073  |
| 30 | Kentucky             | 0.000788918 | -0.0167411   | 0.829858    |  0.215872  |
| 25 | Virginia             | 0.000731319 | -0.0167997   | 0.843439    |  0.198399  |
| 13 | Idaho                | 0.000448908 | -0.0170871   | 0.8502      |  0.189723  |
| 34 | Michigan             | 0.000347234 | -0.0171905   | 0.846214    | -0.194837  |
| 39 | Ohio                 | 0.000323144 | -0.017215    | 0.851524    | -0.188026  |
| 18 | California           | 0.000180932 | -0.0173598   | 0.833919    |  0.21064   |
| 23 | Pennsylvania         | 3.80019e-05 | -0.0175052   | 0.951912    |  0.0605714 |
| 38 | South-Dakota         | 2.92915e-05 | -0.0175141   | 0.764988    | -0.30037   |

In [ ]:
import statsmodels.api as sm

ols_fit_overall=pd.DataFrame()
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
for i in set(merged_df["state"]):
  states_df=merged_df[merged_df["state"]==i]
  listx=calculate_difference_list(states_df["high_amount_spent"].tolist())
  listy=calculate_difference_list(states_df["deaths"].tolist())
  listx = sm.add_constant(listx)
  ols = sm.OLS(listy,listx,missing = 'drop')
  ols_result = ols.fit()
  ols_new_results=ols_result.get_robustcov_results()
  print(ols_new_results.summary())
  states.append(i)
  rsq.append(ols_new_results.rsquared)
  adjrsq.append(ols_new_results.rsquared_adj)
  pval.append(ols_new_results.pvalues[1])
  tval.append(ols_new_results.tvalues[1])

ols_fit_overall["State"]=states
ols_fit_overall["R^2"]=rsq
ols_fit_overall["Adj_R^2"]=adjrsq
ols_fit_overall["pval"]=pval
ols_fit_overall["tval"]=tval
ols_fit_overall.sort_values(by="R^2", ascending = False).to_markdown()

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.017
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                     1.323
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.255
Time:                        03:58:43   Log-Likelihood:                -200.05
No. Observations:                  59   AIC:                             404.1
Df Residuals:                      57   BIC:                             408.2
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2231      0.949      0.235      0.8

'|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |\n|---:|:---------------------|------------:|-------------:|------------:|-----------:|\n| 19 | North-Dakota         | 0.118069    |  0.102596    | 0.000404502 |  3.75838   |\n| 42 | Texas                | 0.107657    |  0.0920021   | 0.0130685   |  2.56221   |\n| 36 | Iowa                 | 0.0814558   |  0.065341    | 0.0219086   |  2.35649   |\n| 33 | Utah                 | 0.0704791   |  0.0541717   | 0.020253    | -2.38848   |\n|  1 | Pennsylvania         | 0.0625605   |  0.0461142   | 0.0221318   | -2.35235   |\n| 17 | Alaska               | 0.0546527   |  0.0380676   | 0.117427    |  1.58973   |\n| 15 | Maine                | 0.0528119   |  0.0361946   | 0.0274964   | -2.26248   |\n| 18 | District-of-Columbia | 0.0491579   |  0.0324764   | 0.0800458   |  1.78218   |\n| 13 | Florida              | 0.0467063   |  0.0299819   | 0.00635227  | -2.83372   |\n| 27 | Maryland             | 0.0453523   

|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |
|---:|:---------------------|------------:|-------------:|------------:|-----------:|
| 26 | North-Dakota         | 0.118069    |  0.102596    | 0.000404502 |  3.75838   |
| 24 | Texas                | 0.107657    |  0.0920021   | 0.0130685   |  2.56221   |
| 45 | Iowa                 | 0.0814558   |  0.065341    | 0.0219086   |  2.35649   |
| 10 | Utah                 | 0.0704791   |  0.0541717   | 0.020253    | -2.38848   |
| 23 | Pennsylvania         | 0.0625605   |  0.0461142   | 0.0221318   | -2.35235   |
|  3 | Alaska               | 0.0546527   |  0.0380676   | 0.117427    |  1.58973   |
| 27 | Maine                | 0.0528119   |  0.0361946   | 0.0274964   | -2.26248   |
| 20 | District-of-Columbia | 0.0491579   |  0.0324764   | 0.0800458   |  1.78218   |
| 42 | Florida              | 0.0467063   |  0.0299819   | 0.00635227  | -2.83372   |
|  0 | Maryland             | 0.0453523   |  0.0286041   | 0.0723392   | -1.83093   |
| 41 | Wisconsin            | 0.0452335   |  0.0284832   | 0.0880761   | -1.73539   |
| 29 | Illinois             | 0.0358022   |  0.0188865   | 0.21293     |  1.25965   |
| 12 | Louisiana            | 0.0295415   |  0.0125159   | 0.154459    | -1.44312   |
| 18 | California           | 0.0251358   |  0.00803294  | 0.172443    |  1.38174   |
|  6 | Kansas               | 0.0233623   |  0.00622832  | 0.210904    |  1.26532   |
| 25 | Virginia             | 0.0216282   |  0.00446378  | 0.260105    | -1.13746   |
| 21 | Arizona              | 0.0195528   |  0.00235202  | 0.112793    |  1.61059   |
| 40 | North-Carolina       | 0.0186494   |  0.00143272  | 0.36316     | -0.916705  |
| 17 | New-Mexico           | 0.0180308   |  0.000803297 | 0.237863    |  1.19287   |
|  9 | Colorado             | 0.0165707   | -0.000682464 | 0.254844    | -1.15025   |
|  2 | South-Carolina       | 0.0146834   | -0.00260289  | 0.211558    | -1.26349   |
|  1 | Massachusetts        | 0.0141354   | -0.00316046  | 0.142625    | -1.48663   |
| 39 | Ohio                 | 0.0126649   | -0.00465678  | 0.440685    | -0.776465  |
| 15 | Oregon               | 0.0104157   | -0.00694548  | 0.416749    | -0.818026  |
| 33 | Montana              | 0.00978653  | -0.00758564  | 0.156678    | -1.43526   |
| 11 | Delaware             | 0.00973841  | -0.0076346   | 0.417508    |  0.816687  |
| 44 | Missouri             | 0.00900515  | -0.00838073  | 0.434165    | -0.787649  |
| 37 | Indiana              | 0.00792578  | -0.00947903  | 0.27712     |  1.09731   |
| 31 | Minnesota            | 0.00758017  | -0.0098307   | 0.562616    | -0.582367  |
| 16 | Mississippi          | 0.00670005  | -0.0107263   | 0.278735    | -1.09359   |
| 30 | Kentucky             | 0.00631576  | -0.0111173   | 0.640754    |  0.469147  |
| 14 | Washington           | 0.00504437  | -0.012411    | 0.490697    | -0.693684  |
| 43 | Georgia              | 0.00495392  | -0.012503    | 0.558115    | -0.58911   |
| 35 | Alabama              | 0.00491326  | -0.0125444   | 0.575094    |  0.563811  |
|  8 | Connecticut          | 0.0045789   | -0.0128846   | 0.600422    |  0.526735  |
|  4 | Nevada               | 0.00331504  | -0.0141707   | 0.560038    |  0.586225  |
| 19 | New-Hampshire        | 0.00224302  | -0.0152615   | 0.717811    | -0.363184  |
| 38 | South-Dakota         | 0.00215042  | -0.0153557   | 0.630706    |  0.48334   |
| 22 | Hawaii               | 0.00177982  | -0.0157328   | 0.775625    | -0.286381  |
|  5 | Tennessee            | 0.00119888  | -0.0163239   | 0.729976    | -0.346858  |
| 36 | Arkansas             | 0.00108992  | -0.0164348   | 0.762437    |  0.303733  |
| 13 | Idaho                | 0.000306616 | -0.0172319   | 0.858014    |  0.179715  |
|  7 | New-Jersey           | 0.00025867  | -0.0172807   | 0.872467    |  0.16125   |
| 34 | Michigan             | 0.000192352 | -0.0173481   | 0.932475    | -0.0851059 |
| 28 | Oklahoma             | 6.78816e-05 | -0.0174748   | 0.963722    |  0.0456832 |
| 32 | West-Virginia        | 4.03492e-05 | -0.0175028   | 0.953009    |  0.059189  |

In [ ]:
import statsmodels.api as sm

ols_fit_overall=pd.DataFrame()
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
for i in set(merged_df["state"]):
  states_df=merged_df[merged_df["state"]==i]
  listx=calculate_difference_list(states_df["high_price_per_gram"].tolist())
  listy=calculate_difference_list(states_df["deaths"].tolist())
  listx = sm.add_constant(listx)
  ols = sm.OLS(listy,listx,missing = 'drop')
  ols_result = ols.fit()
  ols_new_results=ols_result.get_robustcov_results()
  print(ols_new_results.summary())
  states.append(i)
  rsq.append(ols_new_results.rsquared)
  adjrsq.append(ols_new_results.rsquared_adj)
  pval.append(ols_new_results.pvalues[1])
  tval.append(ols_new_results.tvalues[1])

ols_fit_overall["State"]=states
ols_fit_overall["R^2"]=rsq
ols_fit_overall["Adj_R^2"]=adjrsq
ols_fit_overall["pval"]=pval
ols_fit_overall["tval"]=tval
ols_fit_overall.sort_values(by="R^2", ascending = False).to_markdown()

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.021
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     1.657
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.203
Time:                        03:58:44   Log-Likelihood:                -199.93
No. Observations:                  59   AIC:                             403.9
Df Residuals:                      57   BIC:                             408.0
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.2494      0.950      0.263      0.7

'|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |\n|---:|:---------------------|------------:|-------------:|------------:|-----------:|\n| 35 | Hawaii               | 0.118329    |  0.102862    | 0.000519837 |  3.6795    |\n| 21 | Oklahoma             | 0.113956    |  0.0984112   | 0.00479038  |  2.93577   |\n| 36 | Iowa                 | 0.108673    |  0.0930358   | 0.0124217   | -2.58187   |\n| 15 | Maine                | 0.0610461   |  0.0445732   | 0.10763     |  1.63467   |\n|  9 | Nevada               | 0.0582165   |  0.041694    | 0.0651085   |  1.88082   |\n| 33 | Utah                 | 0.0458428   |  0.0291032   | 0.107758    | -1.63406   |\n|  7 | Massachusetts        | 0.0381079   |  0.0212326   | 0.0482414   |  2.01861   |\n| 29 | Delaware             | 0.0354724   |  0.0185509   | 0.139764    |  1.49757   |\n| 11 | Arizona              | 0.026664    |  0.00958795  | 0.0180533   | -2.4348    |\n| 19 | North-Dakota         | 0.024708    

|    | State                |         R^2 |      Adj_R^2 |        pval |       tval |
|---:|:---------------------|------------:|-------------:|------------:|-----------:|
| 22 | Hawaii               | 0.118329    |  0.102862    | 0.000519837 |  3.6795    |
| 28 | Oklahoma             | 0.113956    |  0.0984112   | 0.00479038  |  2.93577   |
| 45 | Iowa                 | 0.108673    |  0.0930358   | 0.0124217   | -2.58187   |
| 27 | Maine                | 0.0610461   |  0.0445732   | 0.10763     |  1.63467   |
|  4 | Nevada               | 0.0582165   |  0.041694    | 0.0651085   |  1.88082   |
| 10 | Utah                 | 0.0458428   |  0.0291032   | 0.107758    | -1.63406   |
|  1 | Massachusetts        | 0.0381079   |  0.0212326   | 0.0482414   |  2.01861   |
| 11 | Delaware             | 0.0354724   |  0.0185509   | 0.139764    |  1.49757   |
| 21 | Arizona              | 0.026664    |  0.00958795  | 0.0180533   | -2.4348    |
| 26 | North-Dakota         | 0.024708    |  0.00759766  | 0.000840777 |  3.52585   |
|  9 | Colorado             | 0.020594    |  0.00341148  | 0.203277    | -1.28706   |
|  5 | Tennessee            | 0.0201958   |  0.0030063   | 0.102736    | -1.65837   |
| 44 | Missouri             | 0.0194156   |  0.00221235  | 0.186657    | -1.33662   |
|  8 | Connecticut          | 0.0187756   |  0.00156111  | 0.0497905   |  2.00436   |
| 43 | Georgia              | 0.017228    | -1.36341e-05 | 0.143534    | -1.48318   |
| 32 | West-Virginia        | 0.0154494   | -0.00182337  | 0.133982    | -1.52023   |
| 33 | Montana              | 0.0146606   | -0.00262605  | 0.295564    |  1.05569   |
| 23 | Pennsylvania         | 0.013742    | -0.00356073  | 0.192468    | -1.31893   |
| 20 | District-of-Columbia | 0.0136845   | -0.00361928  | 0.13034     |  1.5349    |
| 25 | Virginia             | 0.0123021   | -0.00502596  | 0.483898    |  0.70465   |
|  2 | South-Carolina       | 0.0112221   | -0.00612492  | 0.341373    |  0.959473  |
| 14 | Washington           | 0.00815065  | -0.00925021  | 0.455828    | -0.750856  |
| 19 | New-Hampshire        | 0.00766402  | -0.00974538  | 0.0853961   |  1.7506    |
| 24 | Texas                | 0.00641176  | -0.0110196   | 0.239252    | -1.1893    |
| 37 | Indiana              | 0.00616619  | -0.0112695   | 0.462602    | -0.739561  |
| 41 | Wisconsin            | 0.00614089  | -0.0112952   | 0.618307    | -0.500991  |
| 31 | Minnesota            | 0.00584438  | -0.0115969   | 0.639253    | -0.471261  |
| 16 | Mississippi          | 0.00512006  | -0.012334    | 0.588216    |  0.544508  |
| 15 | Oregon               | 0.00388781  | -0.0135878   | 0.481667    | -0.708266  |
|  0 | Maryland             | 0.00346282  | -0.0140203   | 0.12636     |  1.55132   |
|  3 | Alaska               | 0.00268928  | -0.0148074   | 0.582758    |  0.552512  |
| 39 | Ohio                 | 0.00251521  | -0.0149845   | 0.425759    |  0.802219  |
| 34 | Michigan             | 0.00208311  | -0.0154242   | 0.509085    | -0.664445  |
| 38 | South-Dakota         | 0.00190313  | -0.0156073   | 0.753537    | -0.315497  |
| 35 | Alabama              | 0.00179588  | -0.0157165   | 0.722352    |  0.357078  |
| 12 | Louisiana            | 0.00138712  | -0.0161324   | 0.679259    |  0.415605  |
| 29 | Illinois             | 0.00138288  | -0.0161367   | 0.612063    |  0.509939  |
|  6 | Kansas               | 0.00124     | -0.0162821   | 0.641074    | -0.468697  |
| 36 | Arkansas             | 0.00116997  | -0.0163534   | 0.710947    |  0.372439  |
| 30 | Kentucky             | 0.0011454   | -0.0163784   | 0.793982    | -0.262371  |
| 18 | California           | 0.000956317 | -0.0165708   | 0.669539    | -0.429002  |
| 42 | Florida              | 0.000650567 | -0.0168819   | 0.810008    | -0.241534  |
| 13 | Idaho                | 0.000390926 | -0.0171461   | 0.869437    | -0.165116  |
| 40 | North-Carolina       | 4.09101e-05 | -0.0175022   | 0.950009    | -0.0629716 |
|  7 | New-Jersey           | 3.80843e-05 | -0.0175051   | 0.913919    |  0.108577  |
| 17 | New-Mexico           | 3.87163e-06 | -0.0175399   | 0.949968    | -0.0630237 |

In [ ]:
import statsmodels.api as sm

ols_fit_overall=pd.DataFrame()
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
for i in set(merged_df["state"]):
  states_df=merged_df[merged_df["state"]==i]
  listx=calculate_difference_list(states_df["overall_price_per_gram"].tolist())
  listy=calculate_difference_list(states_df["injured"].tolist())
  listx = sm.add_constant(listx)
  ols = sm.OLS(listy,listx,missing = 'drop')
  ols_result = ols.fit()
  ols_new_results=ols_result.get_robustcov_results()
  print(ols_new_results.summary())
  states.append(i)
  rsq.append(ols_new_results.rsquared)
  adjrsq.append(ols_new_results.rsquared_adj)
  pval.append(ols_new_results.pvalues[1])
  tval.append(ols_new_results.tvalues[1])

ols_fit_overall["State"]=states
ols_fit_overall["R^2"]=rsq
ols_fit_overall["Adj_R^2"]=adjrsq
ols_fit_overall["pval"]=pval
ols_fit_overall["tval"]=tval
ols_fit_overall.sort_values(by="R^2", ascending = False).to_markdown()

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.072
Model:                            OLS   Adj. R-squared:                  0.055
Method:                 Least Squares   F-statistic:                     31.41
Date:                Thu, 30 Mar 2023   Prob (F-statistic):           6.33e-07
Time:                        03:58:45   Log-Likelihood:                -212.90
No. Observations:                  59   AIC:                             429.8
Df Residuals:                      57   BIC:                             434.0
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.4599      1.183      0.389      0.6

'|    | State                |         R^2 |      Adj_R^2 |        pval |        tval |\n|---:|:---------------------|------------:|-------------:|------------:|------------:|\n| 17 | Alaska               | 0.137789    |  0.122663    | 1.56953e-07 |  5.97686    |\n| 33 | Utah                 | 0.115574    |  0.100057    | 8.696e-10   |  7.33798    |\n| 26 | Michigan             | 0.0761554   |  0.0599476   | 0.000232623 | -3.92953    |\n|  0 | Colorado             | 0.071543    |  0.0552543   | 6.32851e-07 | -5.60457    |\n| 42 | Texas                | 0.0675464   |  0.0511876   | 0.00983048  | -2.67135    |\n| 24 | Ohio                 | 0.0644754   |  0.0480627   | 8.27103e-05 | -4.24043    |\n| 43 | New-Jersey           | 0.0580755   |  0.0415505   | 0.00317614  | -3.08079    |\n| 40 | Oregon               | 0.0497622   |  0.0330913   | 0.000214666 | -3.95408    |\n| 29 | Delaware             | 0.0464795   |  0.0297511   | 0.0631975   | -1.89478    |\n| 37 | New-Hampshire        | 0

|    | State                |         R^2 |      Adj_R^2 |        pval |        tval |
|---:|:---------------------|------------:|-------------:|------------:|------------:|
|  3 | Alaska               | 0.137789    |  0.122663    | 1.56953e-07 |  5.97686    |
| 10 | Utah                 | 0.115574    |  0.100057    | 8.696e-10   |  7.33798    |
| 34 | Michigan             | 0.0761554   |  0.0599476   | 0.000232623 | -3.92953    |
|  9 | Colorado             | 0.071543    |  0.0552543   | 6.32851e-07 | -5.60457    |
| 24 | Texas                | 0.0675464   |  0.0511876   | 0.00983048  | -2.67135    |
| 39 | Ohio                 | 0.0644754   |  0.0480627   | 8.27103e-05 | -4.24043    |
|  7 | New-Jersey           | 0.0580755   |  0.0415505   | 0.00317614  | -3.08079    |
| 15 | Oregon               | 0.0497622   |  0.0330913   | 0.000214666 | -3.95408    |
| 11 | Delaware             | 0.0464795   |  0.0297511   | 0.0631975   | -1.89478    |
| 19 | New-Hampshire        | 0.0431824   |  0.0263961   | 0.092248    | -1.71246    |
| 21 | Arizona              | 0.0342943   |  0.0173521   | 0.179052    | -1.36042    |
| 13 | Idaho                | 0.03424     |  0.0172968   | 0.182758    |  1.34873    |
|  8 | Connecticut          | 0.0326378   |  0.0156665   | 0.0644974   |  1.88525    |
| 36 | Arkansas             | 0.0318879   |  0.0149035   | 0.0690085   |  1.85336    |
|  2 | South-Carolina       | 0.0297498   |  0.0127279   | 0.0207743   | -2.37816    |
| 38 | South-Dakota         | 0.0275034   |  0.0104421   | 0.000349077 |  3.80434    |
| 31 | Minnesota            | 0.0171456   | -9.74127e-05 | 0.335653    | -0.970996   |
| 12 | Louisiana            | 0.0166285   | -0.000623639 | 0.284082    | -1.08138    |
| 33 | Montana              | 0.0160357   | -0.00122688  | 0.245214    |  1.17416    |
|  5 | Tennessee            | 0.0156908   | -0.00157775  | 0.451976    | -0.757322   |
| 37 | Indiana              | 0.0127873   | -0.00453226  | 0.20123     |  1.29299    |
| 45 | Iowa                 | 0.0123298   | -0.00499773  | 0.408039    |  0.833506   |
| 42 | Florida              | 0.0106146   | -0.00674305  | 0.265172    | -1.12531    |
| 35 | Alabama              | 0.00837763  | -0.00901925  | 0.30571     |  1.03355    |
| 30 | Kentucky             | 0.00735477  | -0.0100601   | 0.584367    |  0.550148   |
| 23 | Pennsylvania         | 0.00691271  | -0.0105099   | 0.334147    |  0.974051   |
| 29 | Illinois             | 0.00686377  | -0.0105597   | 0.115973    | -1.5962     |
| 25 | Virginia             | 0.00630829  | -0.0111249   | 0.532227    |  0.628439   |
| 32 | West-Virginia        | 0.00618005  | -0.0112554   | 0.286508    | -1.07589    |
| 28 | Oklahoma             | 0.00613652  | -0.0112997   | 0.378376    |  0.887806   |
| 44 | Missouri             | 0.00584025  | -0.0116011   | 0.00673988  |  2.81203    |
| 14 | Washington           | 0.00555861  | -0.0118877   | 0.461912    |  0.740706   |
| 16 | Mississippi          | 0.00531208  | -0.0121386   | 0.520581    |  0.646452   |
| 17 | New-Mexico           | 0.00500809  | -0.0124479   | 0.58792     |  0.54494    |
|  0 | Maryland             | 0.00476188  | -0.0126984   | 0.390032    |  0.866163   |
| 26 | North-Dakota         | 0.00457595  | -0.0128876   | 0.142981    |  1.48527    |
| 40 | North-Carolina       | 0.0045633   | -0.0129005   | 0.148383    | -1.46512    |
| 41 | Wisconsin            | 0.00452203  | -0.0129425   | 0.581404    |  0.554503   |
| 20 | District-of-Columbia | 0.00332838  | -0.0141571   | 0.585632    |  0.548292   |
|  1 | Massachusetts        | 0.00276984  | -0.0147254   | 0.726748    |  0.351181   |
|  6 | Kansas               | 0.00264028  | -0.0148573   | 0.67475     | -0.421811   |
| 22 | Hawaii               | 0.00185346  | -0.0156579   | 0.256545    |  1.14609    |
| 43 | Georgia              | 0.00117191  | -0.0163514   | 0.618426    | -0.50082    |
| 18 | California           | 0.000486291 | -0.017049    | 0.798333    |  0.256702   |
| 27 | Maine                | 9.05082e-08 | -0.0175438   | 0.997307    | -0.0033895  |
|  4 | Nevada               | 6.08344e-08 | -0.0175438   | 0.998914    |  0.00136702 |

In [ ]:
merged_df

month  year  overall_amount_spent  overall_grams  \
0       January  2013                  1481          221.0   
1      February  2013                  1600          276.5   
2         March  2013                  2245          330.0   
3         April  2013                  1925          285.0   
4           May  2013                  1945          225.0   
...         ...   ...                   ...            ...   
2755     August  2017                  4320          378.5   
2756  September  2017                  1192           90.0   
2757    October  2017                  2912          380.5   
2758   November  2017                  1802          172.0   
2759   December  2017                  2223          240.0   

      overall_price_per_gram      state  high_amount_spent  high_grams  \
0                  10.109524    Alabama               1201       124.0   
1                   7.309524    Alabama                505        94.5   
2                   7.592857    Alabama               1985       236.0   
3                   8.916667    Alabama               1510       173.0   
4                  11.630952    Alabama               1110        69.5   
...                      ...        ...                ...         ...   
2755               11.677019  Wisconsin               2765       213.5   
2756               13.203297  Wisconsin               1080        77.0   
2757               11.004658  Wisconsin               1976       231.0   
2758               13.667857  Wisconsin               1617       140.0   
2759               10.972689  Wisconsin               1420       147.0   

      high_price_per_gram  low_amount_spent  low_grams  low_price_per_gram  \
0               12.484127               135       12.0           12.500000   
1                8.750000               100       28.0            3.571429   
2                9.803571               100       28.0            3.571429   
3               11.285714               100       28.0            3.571429   
4               17.040816                10        3.5            2.857143   
...                   ...               ...        ...                 ...   
2755            14.206349                60        3.5           17.142857   
2756            14.226190                60        3.5           17.142857   
2757             8.376984                60        3.5           17.142857   
2758            11.550000                60        3.5           17.142857   
2759            10.816327                60        3.5           17.142857   

      medium_amount_spent  medium_grams  medium_price_per_gram  deaths  \
0                     280          97.0               6.547619       0   
1                     995         154.0               6.924603       0   
2                     260          94.0               3.171429       0   
3                     415         112.0               4.178571       0   
4                     825         152.0               7.474490       0   
...                   ...           ...                    ...     ...   
2755                 1555         165.0              10.051020      18   
2756                  112          13.0              12.326531      20   
2757                  936         149.5              12.693878      17   
2758                  185          32.0              15.785714      15   
2759                  803          93.0              11.082143      15   

      injured  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
2755       56  
2756       55  
2757       45  
2758       57  
2759       19  

[2760 rows x 17 columns]

In [ ]:
result = merged_df.groupby(['year', 'month'], as_index=False).mean()

In [ ]:
result

year      month  overall_amount_spent  overall_grams  \
0   2013      April           2212.195652     261.380435   
1   2013     August           2683.282609     303.782609   
2   2013   December           1786.934783     222.369565   
3   2013   February           3053.086957     330.717391   
4   2013    January           3520.608696     389.826087   
5   2013       July           2677.869565     311.130435   
6   2013       June           2637.782609     300.282609   
7   2013      March           2581.956522     296.576087   
8   2013        May           3649.434783     440.750000   
9   2013   November           2415.347826     292.793478   
10  2013    October           2604.586957     330.652174   
11  2013  September           2711.586957     312.369565   
12  2014      April           2127.826087     285.130435   
13  2014     August           3489.130435     406.119565   
14  2014   December           2035.239130     261.695652   
15  2014   February           2794.500000     364.391304   
16  2014    January           1895.869565     236.250000   
17  2014       July           2072.195652     250.271739   
18  2014       June           2103.260870     276.532609   
19  2014      March           2776.565217     362.543478   
20  2014        May           2105.760870     281.184783   
21  2014   November           2228.826087     284.195652   
22  2014    October           2441.260870     289.032609   
23  2014  September           2870.173913     347.271739   
24  2015      April           2277.239130     307.967391   
25  2015     August           2231.913043     210.130435   
26  2015   December           1392.956522     142.021739   
27  2015   February           2060.673913     275.869565   
28  2015    January           1977.152174     262.206522   
29  2015       July           1907.000000     190.771739   
30  2015       June           2035.695652     217.380435   
31  2015      March           2360.717391     319.967391   
32  2015        May           3218.913043     406.271739   
33  2015   November           1641.543478     164.010870   
34  2015    October           1793.956522     170.641304   
35  2015  September           1514.543478     143.239130   
36  2016      April           2009.782609     208.565217   
37  2016     August           2052.130435     196.054348   
38  2016   December           1307.978261     129.641304   
39  2016   February           1985.413043     195.836957   
40  2016    January           1620.195652     170.293478   
41  2016       July           1733.000000     185.673913   
42  2016       June           1486.130435     145.010870   
43  2016      March           2387.760870     258.239130   
44  2016        May           1748.782609     177.250000   
45  2016   November           1761.108696     170.771739   
46  2016    October           1881.673913     179.228261   
47  2016  September           3202.913043     279.989130   
48  2017      April           1485.695652     151.760870   
49  2017     August           2399.391304     267.989130   
50  2017   December           1869.239130     211.119565   
51  2017   February           1575.043478     160.065217   
52  2017    January           1412.739130     142.521739   
53  2017       July           2081.043478     233.869565   
54  2017       June           3091.391304     346.956522   
55  2017      March           1571.434783     162.956522   
56  2017        May           2277.739130     246.478261   
57  2017   November           1942.434783     223.119565   
58  2017    October           1847.565217     215.206522   
59  2017  September           2017.673913     220.478261   

    overall_price_per_gram  high_amount_spent  high_grams  \
0                11.342697        1306.043478  134.239130   
1                11.328900        1571.173913  148.793478   
2                11.119161        1091.304348  110.260870   
3                11.653635        1921.543478  176.434783   
4                11.490912        2063

In [ ]:
import statsmodels.api as sm
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
listx=calculate_difference_list(result["overall_price_per_gram"].tolist())
listy=calculate_difference_list(result["injured"].tolist())
listx = sm.add_constant(listx)
ols = sm.OLS(listy,listx,missing = 'drop')
ols_result = ols.fit()
ols_new_results=ols_result.get_robustcov_results()
print(ols_new_results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.024
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1.052
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.309
Time:                        03:58:45   Log-Likelihood:                -213.80
No. Observations:                  59   AIC:                             431.6
Df Residuals:                      57   BIC:                             435.8
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.8740      1.201      0.728      0.4

In [ ]:
import statsmodels.api as sm
states=[]
rsq=[]
adjrsq=[]
pval=[]
tval=[]
listx=calculate_difference_list(result["overall_amount_spent"].tolist())
listy=calculate_difference_list(result["injured"].tolist())
listx = sm.add_constant(listx)
ols = sm.OLS(listy,listx,missing = 'drop')
ols_result = ols.fit()
ols_new_results=ols_result.get_robustcov_results()
print(ols_new_results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.018
Method:                 Least Squares   F-statistic:                  0.001718
Date:                Thu, 30 Mar 2023   Prob (F-statistic):              0.967
Time:                        03:58:45   Log-Likelihood:                -214.52
No. Observations:                  59   AIC:                             433.0
Df Residuals:                      57   BIC:                             437.2
Df Model:                           1                                         
Covariance Type:                  HC1                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.8819      1.215      0.726      0.4

Granger Causality

In [ ]:
from statsmodels.tsa.stattools import grangercausalitytests

def granger_causality(df, col1, col2,state,maxlag):
    df=df[df["state"]==state]
    data = df[[col1, col2]].values
    results = grangercausalitytests(data, maxlag=maxlag, verbose=False)
    return results

In [ ]:
merged_df

month  year  overall_amount_spent  overall_grams  \
0       January  2013                  1481          221.0   
1      February  2013                  1600          276.5   
2         March  2013                  2245          330.0   
3         April  2013                  1925          285.0   
4           May  2013                  1945          225.0   
...         ...   ...                   ...            ...   
2755     August  2017                  4320          378.5   
2756  September  2017                  1192           90.0   
2757    October  2017                  2912          380.5   
2758   November  2017                  1802          172.0   
2759   December  2017                  2223          240.0   

      overall_price_per_gram      state  high_amount_spent  high_grams  \
0                  10.109524    Alabama               1201       124.0   
1                   7.309524    Alabama                505        94.5   
2                   7.592857    Alabama               1985       236.0   
3                   8.916667    Alabama               1510       173.0   
4                  11.630952    Alabama               1110        69.5   
...                      ...        ...                ...         ...   
2755               11.677019  Wisconsin               2765       213.5   
2756               13.203297  Wisconsin               1080        77.0   
2757               11.004658  Wisconsin               1976       231.0   
2758               13.667857  Wisconsin               1617       140.0   
2759               10.972689  Wisconsin               1420       147.0   

      high_price_per_gram  low_amount_spent  low_grams  low_price_per_gram  \
0               12.484127               135       12.0           12.500000   
1                8.750000               100       28.0            3.571429   
2                9.803571               100       28.0            3.571429   
3               11.285714               100       28.0            3.571429   
4               17.040816                10        3.5            2.857143   
...                   ...               ...        ...                 ...   
2755            14.206349                60        3.5           17.142857   
2756            14.226190                60        3.5           17.142857   
2757             8.376984                60        3.5           17.142857   
2758            11.550000                60        3.5           17.142857   
2759            10.816327                60        3.5           17.142857   

      medium_amount_spent  medium_grams  medium_price_per_gram  deaths  \
0                     280          97.0               6.547619       0   
1                     995         154.0               6.924603       0   
2                     260          94.0               3.171429       0   
3                     415         112.0               4.178571       0   
4                     825         152.0               7.474490       0   
...                   ...           ...                    ...     ...   
2755                 1555         165.0              10.051020      18   
2756                  112          13.0              12.326531      20   
2757                  936         149.5              12.693878      17   
2758                  185          32.0              15.785714      15   
2759                  803          93.0              11.082143      15   

      injured  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
2755       56  
2756       55  
2757       45  
2758       57  
2759       19  

[2760 rows x 17 columns]

In [ ]:
states=set(merged_df["state"].tolist())

In [ ]:
granger=pd.DataFrame()
state=[]
lagl=[]
fval=[]
pval=[]
for i in states:
  results=granger_causality(merged_df, 'overall_amount_spent',"deaths",i,5)
  for lag, result in results.items():
    print(f"Lag {lag}: F-value = {result[0]['params_ftest'][0]:.2f}, p-value = {result[0]['params_ftest'][1]:.2f}")
    state.append(i)
    lagl.append(lag)
    fval.append(result[0]['params_ftest'][0])
    pval.append(result[0]['params_ftest'][1])
granger["state"]=state
granger["lag"]=lagl
granger["fvalue"]=fval
granger["pvalue"]=pval

Lag 1: F-value = 0.04, p-value = 0.84
Lag 2: F-value = 0.75, p-value = 0.48
Lag 3: F-value = 0.67, p-value = 0.57
Lag 4: F-value = 0.51, p-value = 0.73
Lag 5: F-value = 0.28, p-value = 0.92
Lag 1: F-value = 0.46, p-value = 0.50
Lag 2: F-value = 1.10, p-value = 0.34
Lag 3: F-value = 0.90, p-value = 0.45
Lag 4: F-value = 0.86, p-value = 0.50
Lag 5: F-value = 1.11, p-value = 0.37
Lag 1: F-value = 0.00, p-value = 0.99
Lag 2: F-value = 0.08, p-value = 0.92
Lag 3: F-value = 0.39, p-value = 0.76
Lag 4: F-value = 0.29, p-value = 0.89
Lag 5: F-value = 0.34, p-value = 0.89
Lag 1: F-value = 0.17, p-value = 0.68
Lag 2: F-value = 0.07, p-value = 0.93
Lag 3: F-value = 0.45, p-value = 0.72
Lag 4: F-value = 1.11, p-value = 0.36
Lag 5: F-value = 0.97, p-value = 0.44
Lag 1: F-value = 1.82, p-value = 0.18
Lag 2: F-value = 1.96, p-value = 0.15
Lag 3: F-value = 2.08, p-value = 0.11
Lag 4: F-value = 1.48, p-value = 0.22
Lag 5: F-value = 0.98, p-value = 0.44
Lag 1: F-value = 0.67, p-value = 0.42
Lag 2: F-val

In [ ]:
granger

state  lag    fvalue    pvalue
0    Colorado    1  0.041672  0.838986
1    Colorado    2  0.753515  0.475684
2    Colorado    3  0.671336  0.573648
3    Colorado    4  0.509163  0.729213
4    Colorado    5  0.279304  0.922007
..        ...  ...       ...       ...
225  Kentucky    1  0.555102  0.459356
226  Kentucky    2  1.802695  0.174815
227  Kentucky    3  1.000336  0.400472
228  Kentucky    4  0.752918  0.561135
229  Kentucky    5  1.388481  0.247136

[230 rows x 4 columns]

In [ ]:
granger=pd.DataFrame()
state=[]
lagl=[]
fval=[]
pval=[]
for i in states:
  results=granger_causality(merged_df, 'overall_price_per_gram',"deaths",i,5)
  for lag, result in results.items():
    print(f"Lag {lag}: F-value = {result[0]['params_ftest'][0]:.2f}, p-value = {result[0]['params_ftest'][1]:.2f}")
    state.append(i)
    lagl.append(lag)
    fval.append(result[0]['params_ftest'][0])
    pval.append(result[0]['params_ftest'][1])
granger["state"]=state
granger["lag"]=lagl
granger["fvalue"]=fval
granger["pvalue"]=pval

Lag 1: F-value = 0.05, p-value = 0.83
Lag 2: F-value = 0.59, p-value = 0.56
Lag 3: F-value = 0.54, p-value = 0.66
Lag 4: F-value = 0.59, p-value = 0.67
Lag 5: F-value = 3.22, p-value = 0.01
Lag 1: F-value = 0.02, p-value = 0.89
Lag 2: F-value = 0.20, p-value = 0.82
Lag 3: F-value = 0.74, p-value = 0.53
Lag 4: F-value = 0.67, p-value = 0.62
Lag 5: F-value = 0.62, p-value = 0.69
Lag 1: F-value = 8.06, p-value = 0.01
Lag 2: F-value = 3.65, p-value = 0.03
Lag 3: F-value = 3.02, p-value = 0.04
Lag 4: F-value = 2.73, p-value = 0.04
Lag 5: F-value = 2.04, p-value = 0.09
Lag 1: F-value = 0.37, p-value = 0.54
Lag 2: F-value = 0.35, p-value = 0.71
Lag 3: F-value = 0.71, p-value = 0.55
Lag 4: F-value = 0.56, p-value = 0.69
Lag 5: F-value = 0.95, p-value = 0.46
Lag 1: F-value = 4.78, p-value = 0.03
Lag 2: F-value = 2.37, p-value = 0.10
Lag 3: F-value = 1.14, p-value = 0.34
Lag 4: F-value = 0.85, p-value = 0.50
Lag 5: F-value = 0.65, p-value = 0.67
Lag 1: F-value = 0.78, p-value = 0.38
Lag 2: F-val

In [ ]:
granger

state  lag    fvalue    pvalue
0    Colorado    1  0.048734  0.826083
1    Colorado    2  0.589854  0.558006
2    Colorado    3  0.538487  0.658112
3    Colorado    4  0.586255  0.674154
4    Colorado    5  3.223286  0.014493
..        ...  ...       ...       ...
225  Kentucky    1  5.516350  0.022394
226  Kentucky    2  2.312437  0.108930
227  Kentucky    3  2.363197  0.082283
228  Kentucky    4  2.100260  0.095704
229  Kentucky    5  2.878897  0.024702

[230 rows x 4 columns]

diffndiff?